plutus-core/plutus-core/test/Value/Spec.hs

```diff
@@ -38,60 +38,121 @@ prop_packPreservesInvariants :: V.NestedMap -> Property
 prop_packPreservesInvariants = checkInvariants . V.pack
 
 -- | Verifies that @insertCoin@ correctly updates the sizes
-prop_insertCoinBookkeeping :: Value -> ValueAmount -> Property
-prop_insertCoinBookkeeping v (ValueAmount amt) =
+prop_insertCoinBookkeeping :: Value -> V.Quantity -> Property
+prop_insertCoinBookkeeping v quantity =
   forAll (genShortHex (V.totalSize v)) $ \currency ->
     forAll (genShortHex (V.totalSize v)) $ \token ->
-      let BuiltinSuccess v' = V.insertCoin (V.unK currency) (V.unK token) amt v
+      let BuiltinSuccess v' =
+            V.insertCoin (V.unK currency) (V.unK token) (V.unQuantity quantity) v
        in checkBookkeeping v'
 
 -- | Verifies that @insertCoin@ preserves @Value@ invariants
-prop_insertCoinPreservesInvariants :: Value -> ValueAmount -> Property
-prop_insertCoinPreservesInvariants v (ValueAmount amt) =
+prop_insertCoinPreservesInvariants :: Value -> V.Quantity -> Property
+prop_insertCoinPreservesInvariants v quantity =
   forAll (genShortHex (V.totalSize v)) $ \currency ->
     forAll (genShortHex (V.totalSize v)) $ \token ->
-      let BuiltinSuccess v' = V.insertCoin (V.unK currency) (V.unK token) amt v
+      let BuiltinSuccess v' =
+            V.insertCoin (V.unK currency) (V.unK token) (V.unQuantity quantity) v
        in checkInvariants v'
 
 prop_unionCommutative :: Value -> Value -> Property
-prop_unionCommutative v v' = V.unionValue v v' === V.unionValue v' v
+prop_unionCommutative v v' =
+  case (V.unionValue v v', V.unionValue v' v) of
+    (BuiltinSuccess r1, BuiltinSuccess r2) -> r1 === r2
+    (BuiltinFailure{}, BuiltinFailure{})   -> property True
+    _                                      -> property False
 
 prop_unionAssociative :: Value -> Value -> Value -> Property
 prop_unionAssociative v1 v2 v3 =
-  V.unionValue v1 (V.unionValue v2 v3) === V.unionValue (V.unionValue v1 v2) v3
+  let succeeded (BuiltinSuccess _)           = True
+      succeeded (BuiltinSuccessWithLogs _ _) = True
+      succeeded (BuiltinFailure _ _)         = False
+
+      extractValue (BuiltinSuccess v)           = v
+      extractValue (BuiltinSuccessWithLogs _ v) = v
+      extractValue (BuiltinFailure _ _)         = error "extractValue called on BuiltinFailure"
+
+      r23 = V.unionValue v2 v3
+      r12 = V.unionValue v1 v2
+   in case (succeeded r23, succeeded r12) of
+        (True, True) ->
+          -- Both intermediate operations succeeded
+          let v23 = extractValue r23
+              v12 = extractValue r12
+              r1 = V.unionValue v1 v23
+              r2 = V.unionValue v12 v3
+           in case (succeeded r1, succeeded r2) of
+                (True, True)   -> extractValue r1 === extractValue r2
+                (False, False) -> property True
+                -- If one final operation fails and the other succeeds, union is not
+                -- associative due to overflow detection. This is expected and correct.
+                _              -> property ()
```

Comment:
This all seems a tad inconsistent. Why explicitly match on `(False, True)` below and use a `_` here? Why `property True` vs `property ()`?

Anyway, I'm nitpicking again.